### PR TITLE
[BIOMAGE-1567] - Move plots into a common container

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
#### Problem
There are inconsistencies of the visual between pages. We'd like this to be uniform so our platform will look more polished.

#### Solution
Refactor page containers into separate component. Created 2 types of components: 
- MultiTileContainer which allows a page to have multi-tile. This is just a refactor of the current tiling functionality that we have so things are much cleaner.
- SingleTileContainer which use divs with the correct styling to provide a visual of having a tile but strips the required imports to keep page size small.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1567

#### Link to staging deployment URL (or set N/A)
https://ui-agi-lanky-frog.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [X] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [X] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [X] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [X] Relevant Github READMEs updated **or** no GitHub README updates required.
- [X] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

**Review by vickymorrison — APPROVED** [review]

**Comment by kafkasl** [comment]
@aerlaut  what is this new page `/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer` reported by bundle size ?

I don't understand what's the difference between `MultiTileContainer` & `SingleTileContainer`.

File: src/pages/experiments/[experimentId]/plots-and-tables/index.jsx [hunk 1]
```diff
@@ -117,39 +15,9 @@ const PlotsTablesHome = ({ experimentId, experimentData }) => {
...
-                  >
-                    <CardItem item={item} />
-                  </Link>
-                </Card>
-              </Col>
-            ))}
-          </Row>
-        ))}
-      </Space>
+      <SingleTileContainer title='Select Plot'>
```
Comment (kafkasl) [inline src/pages/experiments/[experimentId]/plots-and-tables/index.jsx]: I think the `Select Plot` title is unnecessary. It's just obvious that you have to select a plot and it makes the screen more cluttered IMO. You already have the title `Plots & Tables` at the top

File: src/pages/data-management/index.jsx [hunk 2]
```diff
@@ -112,7 +104,7 @@ const DataManagementPage = () => {
           );
         }
 
-        return (<ProjectDetails width={width} height={height} />);
```
Comment (kafkasl) [inline src/pages/data-management/index.jsx:115]: why are these removed?

File: src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+];
+
+const PlotsTablesContainer = (props) => {
+  const { width, height, experimentId } = props;
+
+  return (
+    <div
+      style={{
+        width: width - (layout.PANEL_PADDING * 2),
+        height: height - (layout.PANEL_HEADING_HEIGHT * 2),
```
Comment (kafkasl) [inline src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx]: I think the plot container should not have a heading at all. It just feels unnecessary to have an extra line saying "Plot". We were trying to remove stuff so I think that just having the "Reset" and "Info" buttons floating on the top right without a header would look much cleaner.

File: src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
+import React from 'react';
```
Comment (kafkasl) [inline src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx]: having this in pages seems to add a new route `/plots-and-tables/PlotsTablesContainer` which I assume it's not what we want. Shouldn't this be placed in `components` ?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 5]
```diff
@@ -316,12 +311,12 @@ const DotPlotPage = (props) => {
         <center>
           <Empty description={(
             <>
-              <Paragraph>
```
Comment (kafkasl) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:319]: have you checked that `Paragraph` styling is the same to `p` in all browsers? I'm concerned you checked that it looked good in your browser but `<Paragraph>` sets the styling for older or different browsers to make it look consistent everywhere.

**Comment by kafkasl** [comment]
in order to try to make PRs smaller maybe you could've split this one and first propose the new tiling components (without doing the migration) so people can check whether they make sense and review quickly, and then just proceed with migrating the other components

**Comment by aerlaut** [comment]
> @aerlaut what is this new page `/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer` reported by bundle size ?
> 
> I don't understand what's the difference between `MultiTileContainer` & `SingleTileContainer`.

Oh yea, this should be under `components`. I'll move it there.

`MultipleTileContainer` is a component that shows multiple tiles in a single page. This is like what we have in Data Management and Data Processing, where we have different panels for different components. We use the `ReactMosaic` library to have this UI. This library also provides additional capabilities such as resizing and re-arranging the position of tiles.

`SingleTileContainer` provides the same "tiling" view like multiple tile, but for a single tile. Because there is only a single component/panel that has to be rendered, interactions provided by the `ReactMosaic` package that is useful when we have multiple panels (e.g. resizing, rearranging) is useless for single tile (we don't resize because the panel will always be 100% or rearrange because that's the only panel). Looking at the generated HTML, the tiling effect of `ReactMosaic` is rendered by a series of `divs` with some styles, so I wrote `SingleTileContainer` to emulate what `ReactMosaic` would render for single tile without importing the libray.

File: src/pages/experiments/[experimentId]/plots-and-tables/index.jsx (hunk 1 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/index.jsx]: Removed

File: src/pages/data-management/index.jsx (hunk 2 above)
Comment (aerlaut) [reply src/pages/data-management/index.jsx:115]: The `width` and `height` properties were used in this component to calculate the height of project details component. I replaced this with css changes to the `ProjectDetails.jsx` and `MultiTIleContainer.jsx` to provide the same effect.

File: src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx (hunk 3 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx]: We're trying to make the look more consistent throughout the platform. In my opinion, these additions make for a better UI and UX.

File: src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx (hunk 4 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/PlotsTablesContainer.jsx]: Yep, my bad. Moved under `components`.

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 5 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:319]: The antd `Paragraph` (https://github.com/ant-design/ant-design/blob/master/components/typography/Paragraph.tsx) replaces paragraphs with `div`s and adds a 1-em margin styling, which is also added for normal (non-antd) paragraph elements. It doesn't have any other internal functionality. I've checked that replacing this tag with paragraphs doesn't change anything.

**Comment by aerlaut** [comment]
> in order to try to make PRs smaller maybe you could've split this one and first propose the new tiling components (without doing the migration) so people can check whether they make sense and review quickly, and then just proceed with migrating the other components

Yea I should've done that. The idea for the tiling came up when I was in the middle of doing the PR. Still, I should have separated it though.

**Comment by ivababukova** [comment]
From UIX perspective, what is the point of saying "Select Plot" at the top of plots and tables?
Screenshot 2022-04-18 at 11 12 38

**Comment by aerlaut** [comment]
> From UIX perspective, what is the point of saying "Select Plot" at the top of plots and tables? Screenshot 2022-04-18 at 11 12 38

That's removed with commit [f0b5434](https://github.com/hms-dbmi-cellenics/ui/pull/672/commits/f0b5434a7ad6328537222d9372aff4bf4aa05e76). The staging link is not yet updated.

**Comment by kafkasl** [comment]
> From UIX perspective, what is the point of saying "Select Plot" at the top of plots and tables? Screenshot 2022-04-18 at 11 12 38

@ivababukova I think it's the same point I raised here for the "Plot" text which just feels unnecessary https://github.com/hms-dbmi-cellenics/ui/pull/672#discussion_r851333093. I would even say that both "Plot" and "Controls" are completely unnecessary. Pretty obvious what they are

image

**Comment by aerlaut** [comment]
> > From UIX perspective, what is the point of saying "Select Plot" at the top of plots and tables? Screenshot 2022-04-18 at 11 12 38
> 
> @ivababukova I think it's the same point I raised here for the "Plot" text which just feels unnecessary [#672 (comment)](https://github.com/hms-dbmi-cellenics/ui/pull/672#discussion_r851333093). I would even say that both "Plot" and "Controls" are completely unnecessary. Pretty obvious what they are
> 
> image

The headers for these components are provided by the `ReactMosaicComponent`. I'll see if I can remove these.

**Comment by aerlaut** [comment]
Looking at the ReactMosaic source code, there is no way to display or not display the "header" bar: https://github.com/nomcopter/react-mosaic/blob/master/src/MosaicWindow.tsx#L70-L72

Using the current implementation, the best that can be done is to leave the headers without text:
image

What do you think @kafkasl  @ivababukova?

**Comment by ivababukova** [comment]
> Looking at the ReactMosaic source code, there is no way to display or not display the "header" bar: https://github.com/nomcopter/react-mosaic/blob/master/src/MosaicWindow.tsx#L70-L72
> 
> Using the current implementation, the best that can be done is to leave the headers without text: image
> 
> What do you think @kafkasl @ivababukova?

I personally don't think it is a big deal. It just looked strange to me, if it is hard to remove it, we can keep it.

**Comment by kafkasl** [comment]
> 

 The controls one empty looks pretty bad. I'd say then to keep "Controls" and remove the "Plot". But I don't have any strong preference anymore if the header can't be removed which would be the best for me.

**Comment by aerlaut** [comment]
> > 
> 
> The controls one empty looks pretty bad. I'd say then to keep "Controls" and remove the "Plot". But I don't have any strong preference anymore if the header can't be removed which would be the best for me.

I think it looks even weirder if one of the tile has title, and the other doesn't have one. So, although it's unnecessary, I think it's better to just keep both.

File: src/components/MultiTileContainer.jsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import { layout } from 'utils/constants';
+
+const renderWindow = (tile, width, height, style) => {
+  if (!tile) return <></>;
+
+  return (
+    <div style={{
+      padding: layout.PANEL_PADDING,
+      overflow: 'auto',
+      ...style,
```
Comment (StefanBabukov) [inline src/components/MultiTileContainer.jsx:15]: why do you need to spread the style at the end?

File: src/components/plots/PlotContainer.jsx [hunk 7]
```diff
@@ -36,42 +52,51 @@ const PlotContainer = (props) => {
...
+      setResetDisabled(true);
       return;
     }
 
-    setEnableReset(true);
+    setResetDisabled(false);
   }, [config]);
 
   const onClickReset = (event) => {
     event.stopPropagation();
```
Comment (StefanBabukov) [inline src/components/plots/PlotContainer.jsx]: can you check if we need to stop event propagation here? I think it might not be needed?

File: src/components/plots/PlotContainer.jsx [hunk 8]
```diff
@@ -84,12 +109,40 @@ const PlotContainer = (props) => {
     </Space>
   );
 
+  const TILE_MAP = {
+    [PLOT]: {
+      toolbarControls: renderPlotToolbarControls(),
+      component: () => <center>{children}</center>,
```
Comment (StefanBabukov) [inline src/components/plots/PlotContainer.jsx]: Do we need this `center` here? Also I read that center is removed from HTML5 and instead you should use `text-align: center` css property if you actually need it - https://www.techonthenet.com/html/elements/center_tag.php#:~:text=Description,as%20the%20element.

File: src/components/plots/PlotResetButton.jsx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  const checkConfigEquality = (currentConfig, initialConfig) => {
+    const ignoredFields = {
+      // config fields that are set dynamically on component render
+      // should not be compared to their initial values
+      frequency: ['proportionGrouping', 'xAxisGrouping'],
+      embeddingCategorical: ['axes'],
+      embeddingContinuous: ['shownGene', 'axes'],
+      violin: ['shownGene'],
+      markerHeatmap: ['selectedGenes', 'groupedTracks'],
+      DotPlot: ['selectedGenes'],
```
Comment (StefanBabukov) [inline src/components/plots/PlotResetButton.jsx]: If we change the plotUuid of a plot this will not work, can you check if you can add the plot names dynamically? Do we have a file with the plot names?

File: src/components/plots/PlotContainer.jsx [hunk 10]
```diff
@@ -36,42 +52,51 @@ const PlotContainer = (props) => {
...
+      {showReset ? (
         <Button
           key='reset'
           type='primary'
           size='small'
           onClick={onClickReset}
-          disabled={!enableReset}
+          disabled={resetDisabled}
         >
           Reset
```
Comment (StefanBabukov) [inline src/components/plots/PlotContainer.jsx:100]: Can you use the ResetButton component you created here and remove all this code including the functions and variables working with it?

File: src/components/plots/PlotsTablesContainer.jsx [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+import { plotNames, layout } from 'utils/constants';
+import heatmap from '../../../public/static/media/heatmap.png';
+import embeddingContinuous from '../../../public/static/media/embeddingContinuous.png';
+import embeddingCategorical from '../../../public/static/media/embeddingCategorical.png';
+import violin from '../../../public/static/media/violin.png';
+import dotPlot from '../../../public/static/media/dotplot.png';
+import volcano from '../../../public/static/media/volcano.png';
+import frequency from '../../../public/static/media/frequency.png';
+import markerHeatmap from '../../../public/static/media/marker_heatmap.png';
```
Comment (StefanBabukov) [inline src/components/plots/PlotsTablesContainer.jsx:16]: change paths to absolute?

File: src/components/plots/styling/violin/ViolinControls.jsx [hunk 12]
```diff
@@ -59,9 +59,14 @@ const ViolinControls = (props) => {
...
 ViolinControls.propTypes = {
-  config: PropTypes.object.isRequired,
+  config: PropTypes.object,
   onUpdate: PropTypes.func.isRequired,
   setSearchedGene: PropTypes.func.isRequired,
   cellSets: PropTypes.object.isRequired,
 };
+
+ViolinControls.defaultProps = {
+  config: null,
```
Comment (StefanBabukov) [inline src/components/plots/styling/violin/ViolinControls.jsx]: What is the use of the controls if config can be null? Why dont we enforce it like before?

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx [hunk 13]
```diff
@@ -27,40 +24,38 @@ import { plotNames } from 'utils/constants';
...
-    },
-    geneNamesFilter: null,
-    sorter: { field: PROPERTIES[0], columnKey: PROPERTIES[0], order: 'descend' },
-  };
-
   const [searchedGene, setSearchedGene] = useState();
 
   useEffect(() => {
-    if (!_.isNil(config?.shownGene) && !searchedGene) {
+    dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx]: I think you should load this only if redux is empty, right? Otherwise we do redundant calls

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 14]
```diff
@@ -214,106 +209,37 @@ const MarkerHeatmap = ({ experimentId }) => {
...
+    spec.description = 'Marker heatmap';
 
-    const newVegaSpec = {
-      ...spec,
-      axes: [...spec.axes, ...displayLabels()],
-      data: spec.data.map((datum) => ({
-        ...datum,
-        values: data[datum.name],
-      })),
+    const extraMarks = {
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:211]: I feel like this stuff should happen inside the generateSpec not here? Also in the other heatmap i saw a similar thing for the axes

File: src/pages/experiments/[experimentId]/plots-and-tables/violin/index.jsx [hunk 15]
```diff
@@ -90,40 +86,39 @@ const ViolinIndex = ({ experimentId }) => {
     />
   );
 
+  const renderPlot = () => {
+    if (!config) {
+      return (
+        <center>
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/violin/index.jsx:92]: We dont need this center I think?

File: src/components/plots/PlotResetButton.jsx (hunk 9 above)
Comment (aerlaut) [reply src/components/plots/PlotResetButton.jsx]: This component is a carry over from a previous commit. This is not used anymore in favor of a direct button as implemented in `PlotContainer.jsx`. I'm deleting this,

File: src/components/plots/PlotContainer.jsx (hunk 8 above)
Comment (aerlaut) [reply src/components/plots/PlotContainer.jsx]: Replaced this by adding styles to the containing div.

**Comment by ivababukova** [comment]
> > in order to try to make PRs smaller maybe you could've split this one and first propose the new tiling components (without doing the migration) so people can check whether they make sense and review quickly, and then just proceed with migrating the other components
> 
> Yea I should've done that. The idea for the tiling came up when I was in the middle of doing the PR. Still, I should have separated it though.

Separating this PR out would have been amazing

File: src/components/plots/PlotsTablesContainer.jsx (hunk 11 above)
Comment (aerlaut) [reply src/components/plots/PlotsTablesContainer.jsx:16]: These can't be changed to the absolute path because they lay outside the `src` folder. We'd have to add a new path for these files. I don't think it's worth it because we don't use it a lot, so I'd leave this.

File: src/components/plots/PlotContainer.jsx (hunk 10 above)
Comment (aerlaut) [reply src/components/plots/PlotContainer.jsx:100]: The `ResetButton` was a carry over from a previous commit, I'm using this instead of that

File: src/components/plots/styling/violin/ViolinControls.jsx (hunk 12 above)
Comment (aerlaut) [reply src/components/plots/styling/violin/ViolinControls.jsx]: changed to be required

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx (hunk 13 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx]: hmm... yea. I'll put guard clauses around these `loadPlotConfig` calls.

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx (hunk 14 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:211]: The generateHeatmapSpec functionality is used for both the marker heatmap and the heatmap plot. What differentiates them is just these markings. I don't think passing the plot type to `generateHeatmapSpec` and using it inside the function to decided what to add is a good way to differentiate between the two. That's why I take them out and put them here.

What I can do to make it look better is maybe to wrap it inside a wrapper function that adds these settings to the output of `generateHeatmapSpec`?

File: src/pages/experiments/[experimentId]/plots-and-tables/violin/index.jsx (hunk 15 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/violin/index.jsx:92]: removed

File: src/components/plots/PlotContainer.jsx (hunk 7 above)
Comment (aerlaut) [reply src/components/plots/PlotContainer.jsx]: Yep, it's not needed. Removed. I think this was also a carry over from a previous PR,

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx (hunk 14 above)
Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:211]: I think that just complicates things more. Maybe just keep it like this?

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx [hunk 16]
```diff
@@ -31,8 +31,8 @@ const EmbeddingCategoricalPage = ({ experimentId }) => {
   const hierarchy = useSelector(getCellSetsHierarchy());
 
   useEffect(() => {
+    if (!config) dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
     dispatch(loadCellSets(experimentId));
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx:35]: Add a check for the cellsets too?

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx [hunk 17]
```diff
@@ -50,7 +50,7 @@ const ContinuousEmbeddingPage = ({ experimentId }) => {
   const [searchedGene, setSearchedGene] = useState();
 
   useEffect(() => {
-    dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
+    if (!config) dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
     dispatch(loadCellSets(experimentId));
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx:54]: Add a check for the cellsets too?

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx (hunk 16 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx:35]: `loadCellSets` has [an internal check](https://github.com/hms-dbmi-cellenics/ui/blob/master/src/redux/actions/cellSets/loadCellSets.js#L13-L15) which would not fire the request if cell set exists, so we don't have to explicitly guard this

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx (hunk 17 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx:54]: `loadCellSets` has [an internal check](https://github.com/hms-dbmi-cellenics/ui/blob/master/src/redux/actions/cellSets/loadCellSets.js#L13-L15) which would not fire the request if cell set exists, so we don't have to explicitly guard this

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx (hunk 14 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:211]: I'll create the wrapper function to make it easier to read.

**Review by StefanBabukov — APPROVED** [review]
lgtm
